src/Features/CSharp/Portable/ConvertLinq/CSharpConvertLinqQueryToForEachProvider.cs

```diff
@@ -0,0 +1,608 @@
+                    if (TryConvertQueryExpression(source, internalNodeMethod, out var convertedFromQueryExpression))
+                    {
+                        if (TryGetTypeSyntax(source, out var typeSyntax))
+                        {
+                            var localFunctionDeclaration = SyntaxFactory.LocalFunctionStatement(
+                                modifiers: default,
+                                returnType: typeSyntax.WithAdditionalAnnotations(Simplifier.SpecialTypeAnnotation),
+                                identifier: SyntaxFactory.Identifier(localFunctionName),
+                                typeParameterList: null,
+                                parameterList: SyntaxFactory.ParameterList(),
+                                constraintClauses: default,
+                                body: SyntaxFactory.Block(
+                                    SyntaxFactory.Token(
+                                        SyntaxFactory.TriviaList(),
+                                        SyntaxKind.OpenBraceToken,
+                                        SyntaxFactory.TriviaList(SyntaxFactory.EndOfLine(Environment.NewLine))),
+                                    SyntaxFactory.SingletonList(convertedFromQueryExpression),
+                                    SyntaxFactory.Token(SyntaxKind.CloseBraceToken)),
+                                expressionBody: null);
+
+                            var localFunctionInvocation = SyntaxFactory.InvocationExpression(SyntaxFactory.IdentifierName(localFunctionName));
+                            SyntaxNode newParentExpressionStatement = parentStatement.ReplaceNode(source, localFunctionInvocation);
+                            documentUpdate = new DocumentUpdate(parentStatement, new[] { localFunctionDeclaration, newParentExpressionStatement });
+                            return true;
+                        }
+                    }
+                }
+
+                documentUpdate = default;
+                return false;
+            }
+
+            private bool TryGetTypeSyntax(ExpressionSyntax expression, out TypeSyntax typeSyntax)
+            {
+                var typeSymbol = _semanticModel.GetTypeInfo(expression, _cancellationToken).Type;
+                if (typeSymbol.TypeKind == TypeKind.Error ||
+                    typeSymbol.ContainsAnonymousType())
+                {
+                    typeSyntax = default;
+                    return false;
+                }
+
+                typeSyntax = typeSymbol.GenerateTypeSyntax();
+                return true;
+            }
+
+            private string GetFreeSymbolName(string prefix, int position)
+                => Shared.Utilities.NameGenerator.GenerateUniqueName(
+                    prefix,
+                    _semanticModel.LookupSymbols(position).Select(symbol => symbol.Name).ToImmutableHashSet(),
+                    StringComparer.CurrentCulture);
```

Comment:
note: you may want to merge in my for-to-foreach branch.  There's a helper to do this for you.  It will also ensure you don't use names that may be declared in nested scopes.

Reply:
Thank you! Will definitely do this but as a separate PR. I have a time restriction for this PR, and prefer to avoid extra merge complications.

added to [https://github.com/dotnet/roslyn/issues/25639](https://github.com/dotnet/roslyn/issues/25639)

---
In reply to: [176297961](https://github.com/dotnet/roslyn/pull/25362#discussion_r176297961) [](ancestors = 176297961)